list.c

```diff
@@ -0,0 +1,96 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/wait.h>
+
+// GOAL: implement the following command :  clear & ls -l | tee text.txt & mv text.txt test.txt
+int main(int argc, char** argv)
+{
+    pid_t pclear_id, pls_id, ptee_id, pmv_id; // create all process id
+    int mypipe[2];
+
+    // Create pipe
+    pipe(mypipe);
+
+    if((pclear_id = fork()) < 0) {
+        // error forking child
+        perror("Error forking child");
+        return 1;
+    } else if(pclear_id == 0) {
+        // Child process
+
+        // No need for pipe here
+        if(close(mypipe[0]) == -1) {
+            perror("Error closing pipe");
+            return 1;
+        }
+
+        if(close(mypipe[1]) == -1) {
+            perror("Error closing pipe");
+            return 1;
+        }
+
+        // Executes the clear command
+        execlp("clear", "clear", NULL);
+    } else {
+        // Parent process
+        if((pls_id = fork()) < 0) { // Error checking
+            perror("Error forking child");
+            return 1;
+        } else if(pls_id == 0) {
+            // No need to read from processes
+            if(close(mypipe[0]) == -1) {
+                perror("Error closing pipe");
+                return 1;
+            }
+
+            dup2(mypipe[1], 1);
+            execlp("ls", "ls", "-l", NULL);
+
+            if(close(mypipe[1]) == -1) {
+                perror("Error closing pipe");
+                return 1;
+            }
+        } else {
+            if((ptee_id = fork()) < 0) {
+                perror("Error forking child");
+                return 1;
+            } else if(ptee_id == 0) {
+                if(close(mypipe[1]) == -1) {
+                    perror("Error closing pipe");
+                    return 1;
+                }
+
+                dup2(mypipe[0], 0);
+                execlp("tee", "tee", "Dir0/t1.txt", NULL);
+
+                if(close(mypipe[0]) == -1) {
+                    perror("Error closing pipe");
+                    return 1;
+                }
+            } else {
+                if((pmv_id = fork()) < 0) {
+                    perror("Error forking child");
+                    return 1;
+                } else if(pmv_id == 0) {
+                    
+                    close(mypipe[0]);
+                    close(mypipe[1]);
+
+                    execlp("mv", "mv", "Dir0/t1.txt", "Dir0/tree.txt", NULL);

```

Comment:
- After you remove the lock, will it cause race condition? For example, `execlp("mv", "mv", "Dir0/t1.txt", "Dir0/tree.txt", NULL);` execute before `execlp("tee", "tee", "Dir0/t1.txt", NULL);`

Reply:
![image](https://user-images.githubusercontent.com/38193373/79678244-4f789900-81c7-11ea-967e-d2b75cf3ec99.png)
This is what I mean by race condition

Reply:
![image](https://user-images.githubusercontent.com/38193373/79678268-89499f80-81c7-11ea-8b0b-232585548c21.png)
And it doesn't always happen
\

Reply:
- Use `wait(NULL)` in parent to wait for child1 finish then fork child2? 
- The other solution can be fork child2 in side of  child1 instead of parent.

Reply:
I tried doing that i got so many problems, so I solved it by using more pipes

Reply:
```
    if((pls_id = fork()) < 0) { // Error checking
        perror("Error forking child");
        return 1;
    } else if(pls_id == 0) {
        execlp();     //child1
    } 

    wait(NULL);
    if((ptee_id = fork()) < 0) {
        perror("Error forking child");
        return 1;
    } else if(ptee_id == 0){
         execlp();     //child2
    }
        
    wait(NULL); 
    printf("child%d\n",4);
    if((pmv_id = fork()) < 0) {
        perror("Error forking child");
        return 1;
    } else if(pmv_id == 0) {
        execlp();     //child3
    }

    wait(NULL);   
```
Have you try something like this, the logic is more clear